src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs

```diff
@@ -20,20 +20,31 @@ internal class VisualStudioExperimentationService : ForegroundThreadAffinitizedO
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
         public VisualStudioExperimentationService(IThreadingContext threadingContext, SVsServiceProvider serviceProvider)
-            : base(threadingContext, assertIsForeground: true)
+            : base(threadingContext)
         {
-            try
+            object experimentationServiceOpt = null;
+            MethodInfo isCachedFlightEnabledInfo = null;
+
+            threadingContext.JoinableTaskFactory.Run(async () =>
             {
-                _experimentationServiceOpt = serviceProvider.GetService(typeof(SVsExperimentationService));
-                if (_experimentationServiceOpt != null)
+                await threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
```

Comment:
dumb question.  how is ths better?  Won't this still block the UI thread while this is happening?

Reply:
> dumb question. how is ths better? Won't this still block the UI thread while this is happening?

In the event this type is constructed on the UI thread, the new code is the same as the old code.

In the event this type is constructed on a background thread, the new code will block the background thread while a UI thread operation completes, rather than crash altogether.

Reply:
Let me rephrase :)

Because we are still just blocking the UI thread, this doesn't seem like much of an improvement.  I'd prefer that since this is going ot block the UI thread, we stick with what we have, and we make whatever component is calling into this acquire the service at the appropriate time.

That said, my ideal would simply be that this is just all async. 

--

However, i'm guessing since this is dev16 this is being done to quickly fix an issue.  So i wouldn't block things here.  But it seems like we should really be fixing this up for real outside of dev16 :)

Reply:
> this doesn't seem like much of an improvement

For the impacted scenario, it's the difference between roslyn working, and roslyn not working at all. 😄 

> we make whatever component is calling into this acquire the service at the appropriate time.

MEF does not guarantee the thread on which objects are initialized. The initializer should be caller-thread-agnostic, either by using JTF or by avoiding code that requires the main thread altogether.

Reply:
No worries.  Doing that later is totally ok with me :)